### PR TITLE
Un-deprecate `ygot.GoStruct`, revert `ygot.ValidatedGoStruct` to pre-v0.18.1

### PR DESCRIPTION
**Pull request by wenovus**
This reverts the [backwards-incompatibility in v0.18.1](https://github.com/openconfig/ygot/releases/tag/v0.18.1) without reverting #643.

**This PR proposes in the near future to deprecate `ValidatedGoStruct` and instead use https://github.com/openconfig/ygot/issues/663 to access all functions currently in `ValidatedGoStruct`.**

The idea is to revert the `ValidatedGoStruct` definition in order to preserve backwards-compatibility and thus break the usage of `ValidatedGoStruct` only for those minority of users using a different name for Validate(). After #663 is rolled out, then these users will no longer be discriminated and a standard API will be used regardless of generation flags.

In order to allow these minority of users access to ygot/ytypes' helper functions, `ygot.GoStruct` has been un-deprecated and used throughout ygot, and a new internal interface containing `ΛValidate()` is used whenever necessary. There is a dearth of internal functions in ygot that actually requires calling these methods so this conversion was pretty hassle-free.

Un-deprecating `ygot.GoStruct` is necessary in order to allow those using a different name for `Validate()` to access ygot/ytypes helpers. This is because their generated structs do not satisfy the `ValidatedGoStruct` interface and hence it cannot be used as the input parameter type. `ygot.GoStruct` is the only other available interface.

### CODE REVIEW

**Comment by robshakir** [comment]
This makes sense to me, thanks @wenovus. I think this keeps things cleaner, given that we know that we have ~0 users that are using a name other than `Validate`.

File: ygot/types.go [hunk 1]
```diff
@@ -58,6 +56,17 @@ type ValidatedGoStruct interface {
 	ΛBelongingModule() string
 }
 
+// validatedGoStruct is an interface used for validating GoStructs.
+type validatedGoStruct interface {
```
Comment (robshakir) [inline ygot/types.go:60]: Essentially, this interface is internal-only such that it is *always* implemented by a generated struct, right?

**Review by robshakir — APPROVED** [review]

File: ygot/types.go (hunk 1 above)
Comment (wenovus) [reply ygot/types.go:60]: Yes, `ΛValidate` and `IsYANGGoStruct()` are always generated.

Comment (robshakir) [reply ygot/types.go:60]: Great, thanks -- can we note this in the description?

Comment (wenovus) [reply ygot/types.go:60]:
```
 59 // validatedGoStruct is an interface used for validating GoStructs.                                                                                                                                                                                                                                                                                    
 60 // This interface is implemented by all Go structs (YANG container or lists),                                                                                                                                                                                                                                                                          
 61 // regardless of generation flag.                                                                                                                                                                                                                                                                                                                      
```
Done.

**Comment by wenovus** [comment]
So this still is a backwards-incompatible change because `ytypes.UnmarshalFunc` and `ytypes.MergeStructs` now take in `ygot.GoStruct` instead of `ygot.ValidatedGoStruct` which may break existing code that expect a parameter or a return value of `ValidatedGoStruct`, although I believe these changes are what we want (since we want to deprecate this interface), and are tolerable enough.

https://github.com/openconfig/ygot/releases/tag/v0.20.0

**Comment by robshakir** [comment]
I agree, this seems much more tolerable than needing to have users change their `.Validate()` calls -- I expect the number of places that folks are casting to `ValidatedGoStruct`.

**Pull request merged**